// routing-controllers/metadata/ParamMetadata.js
"use strict";
var ParamMetadata = (function () {
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    function ParamMetadata(actionMetadata, args) {
        this.actionMetadata = actionMetadata;
        this.target = args.target;
        this.method = args.method;
        this.reflectedType = args.reflectedType;
        if (args.index !== undefined)
            this.index = args.index;
        if (args.type)
            this.type = args.type;
        if (args.name)
            this.name = args.name;
        if (args.format)
            this.format = args.format;
        if (args.parseJson)
            this.parseJson = args.parseJson;
        if (args.isRequired)
            this.isRequired = args.isRequired;
        if (args.transform)
            this.transform = args.transform;
        if (args.classTransformOptions)
            this.classTransformOptions = args.classTransformOptions;
        this.extraOptions = args.extraOptions;
    }
    return ParamMetadata;
}());
exports.ParamMetadata = ParamMetadata;
