// routing-controllers/decorator/controllers.js
"use strict";
var index_1 = require("../index");
/**
 * Defines a class as a controller. All methods with special decorators will be registered as controller actions.
 * Controller actions are executed when request to their routes comes.
 *
 * @param baseRoute Extra path you can apply as a base route to all controller actions
 */
function Controller(baseRoute) {
    return function (object) {
        var metadata = {
            route: baseRoute,
            target: object,
            type: "default"
        };
        index_1.defaultMetadataArgsStorage().controllers.push(metadata);
    };
}
exports.Controller = Controller;
/**
 * Defines a class as a JSON controller. If JSON controller is used, then all controller actions will return
 * a serialized json data, and its response content-type always will be application/json.
 *
 * @param baseRoute Extra path you can apply as a base route to all controller actions
 */
function JsonController(baseRoute) {
    return function (object) {
        var metadata = {
            route: baseRoute,
            target: object,
            type: "json"
        };
        index_1.defaultMetadataArgsStorage().controllers.push(metadata);
    };
}
exports.JsonController = JsonController;
