// routing-controllers/decorator/methods.js
"use strict";
var index_1 = require("../index");
var ActionTypes_1 = require("../metadata/types/ActionTypes");
function Get(route) {
    return function (object, methodName) {
        var metadata = {
            route: route,
            target: object.constructor,
            method: methodName,
            type: ActionTypes_1.ActionTypes.GET
        };
        index_1.defaultMetadataArgsStorage().actions.push(metadata);
    };
}
exports.Get = Get;
function Post(route) {
    return function (object, methodName) {
        var metadata = {
            route: route,
            target: object.constructor,
            method: methodName,
            type: ActionTypes_1.ActionTypes.POST
        };
        index_1.defaultMetadataArgsStorage().actions.push(metadata);
    };
}
exports.Post = Post;
function Put(route) {
    return function (object, methodName) {
        var metadata = {
            route: route,
            target: object.constructor,
            method: methodName,
            type: ActionTypes_1.ActionTypes.PUT
        };
        index_1.defaultMetadataArgsStorage().actions.push(metadata);
    };
}
exports.Put = Put;
function Patch(route) {
    return function (object, methodName) {
        var metadata = {
            route: route,
            target: object.constructor,
            method: methodName,
            type: ActionTypes_1.ActionTypes.PATCH
        };
        index_1.defaultMetadataArgsStorage().actions.push(metadata);
    };
}
exports.Patch = Patch;
function Delete(route) {
    return function (object, methodName) {
        var metadata = {
            route: route,
            target: object.constructor,
            method: methodName,
            type: ActionTypes_1.ActionTypes.DELETE
        };
        index_1.defaultMetadataArgsStorage().actions.push(metadata);
    };
}
exports.Delete = Delete;
function Head(route) {
    return function (object, methodName) {
        var metadata = {
            route: route,
            target: object.constructor,
            method: methodName,
            type: ActionTypes_1.ActionTypes.HEAD
        };
        index_1.defaultMetadataArgsStorage().actions.push(metadata);
    };
}
exports.Head = Head;
function Method(method, route) {
    return function (object, methodName) {
        var metadata = {
            route: route,
            target: object.constructor,
            method: methodName,
            type: method
        };
        index_1.defaultMetadataArgsStorage().actions.push(metadata);
    };
}
exports.Method = Method;
