// routing-controllers/metadata-builder/MetadataArgsStorage.js
"use strict";
/**
 * Storage all metadatas read from decorators.
 */
var MetadataArgsStorage = (function () {
    function MetadataArgsStorage() {
        // -------------------------------------------------------------------------
        // Properties
        // -------------------------------------------------------------------------
        this.controllers = [];
        this.middlewares = [];
        this.interceptors = [];
        this.uses = [];
        this.useInterceptors = [];
        this.actions = [];
        this.params = [];
        this.responseHandlers = [];
    }
    // -------------------------------------------------------------------------
    // Public Methods
    // -------------------------------------------------------------------------
    MetadataArgsStorage.prototype.findMiddlewareMetadatasForClasses = function (classes) {
        return this.middlewares.filter(function (ctrl) {
            return classes.filter(function (cls) { return ctrl.target === cls; }).length > 0;
        });
    };
    MetadataArgsStorage.prototype.findInterceptorMetadatasForClasses = function (classes) {
        return this.interceptors.filter(function (ctrl) {
            return classes.filter(function (cls) { return ctrl.target === cls; }).length > 0;
        });
    };
    MetadataArgsStorage.prototype.findControllerMetadatasForClasses = function (classes) {
        return this.controllers.filter(function (ctrl) {
            return classes.filter(function (cls) { return ctrl.target === cls; }).length > 0;
        });
    };
    MetadataArgsStorage.prototype.findActionsWithTarget = function (target) {
        return this.actions.filter(function (action) { return action.target === target; });
    };
    MetadataArgsStorage.prototype.findUsesWithTargetAndMethod = function (target, methodName) {
        return this.uses.filter(function (use) {
            return use.target === target && use.method === methodName;
        });
    };
    MetadataArgsStorage.prototype.findUseInterceptorWithTargetAndMethod = function (target, methodName) {
        return this.useInterceptors.filter(function (interceptor) {
            return interceptor.target === target && interceptor.method === methodName;
        });
    };
    MetadataArgsStorage.prototype.findParamsWithTargetAndMethod = function (target, methodName) {
        return this.params.filter(function (param) {
            return param.target === target && param.method === methodName;
        });
    };
    MetadataArgsStorage.prototype.findResponseHandlersWithTargetAndMethod = function (target, methodName) {
        return this.responseHandlers.filter(function (property) {
            return property.target === target && property.method === methodName;
        });
    };
    /**
     * Removes all saved metadata.
     */
    MetadataArgsStorage.prototype.reset = function () {
        this.controllers = [];
        this.middlewares = [];
        this.interceptors = [];
        this.uses = [];
        this.useInterceptors = [];
        this.actions = [];
        this.params = [];
        this.responseHandlers = [];
    };
    return MetadataArgsStorage;
}());
exports.MetadataArgsStorage = MetadataArgsStorage;
